File: MoveTrack-iOS/widowmaker/MoveTrack/Views/Nexus/AgentMessageParser.swift
//
//  AgentMessageParser.swift
//  Nexus Moves
//
//  Parses the agent's inline `[BUTTONS]…[/BUTTONS]` blocks into tappable buttons
//  and strips them (plus [IMG:…] tokens) from the displayed prose. Mirrors the
//  web contract in movetrack-app/src/components/nexus/NexusChat.vue:
//
//    [BUTTONS]
//    Label|message to send|actions
//    [/BUTTONS]
//
//  actions ∈ {send, prefill, camera} (optional; inferred when omitted).
//

import Foundation

enum AgentButtonAction {
    case send     // send `message` as a user message immediately
    case prefill  // drop `message` into the input for the user to complete
    case camera   // open the camera, using `message` as the caption
}

struct AgentButton: Identifiable {
    let id = UUID()
    let label: String
    let message: String
    let actions: [AgentButtonAction]

    /// The single action we act on (priority: prefill > camera > send), matching
    /// the web's handleButtonClick precedence.
    var primaryAction: AgentButtonAction {
        if actions.contains(where: { if case .prefill = $0 { return true }; return false }) { return .prefill }
        if actions.contains(where: { if case .camera = $0 { return true }; return false }) { return .camera }
        return .send
    }
}

struct ParsedAgentMessage {
    let prose: String
    let buttons: [AgentButton]
}

enum AgentMessageParser {
    static func parse(_ content: String) -> ParsedAgentMessage {
        ParsedAgentMessage(prose: prose(content), buttons: buttons(content))
    }

    // MARK: - Buttons

    static func buttons(_ content: String) -> [AgentButton] {
        guard let inner = firstGroup(pattern: "\\[buttons\\]([\\s\\S]*?)\\[/buttons\\]", in: content) else {
            return []
        }
        return inner.split(separator: "\n").compactMap { rawLine -> AgentButton? in
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard line.contains("|") else { return nil }
            let parts = line.components(separatedBy: "|")
            let label = parts[0].trimmingCharacters(in: .whitespaces)
            let lastPart = (parts.last ?? "").trimmingCharacters(in: .whitespaces)
            let looksLikeActions = parts.count >= 3
                && lastPart.count < 30
                && lastPart.range(of: "^[a-z, ]+$", options: [.regularExpression, .caseInsensitive]) != nil

            let messageParts = looksLikeActions ? Array(parts[1..<(parts.count - 1)]) : Array(parts[1...])
            let message = messageParts.joined(separator: "|").trimmingCharacters(in: .whitespaces)
            let explicit = looksLikeActions ? parseActions(lastPart) : []
            let actions = explicit.isEmpty ? inferActions(message) : explicit

            guard !label.isEmpty, !message.isEmpty else { return nil }
            return AgentButton(label: label, message: message, actions: actions)
        }
    }

    private static func parseActions(_ raw: String) -> [AgentButtonAction] {
        raw.lowercased().split(whereSeparator: { $0 == "," || $0 == " " }).compactMap { token in
            switch token.replacingOccurrences(of: "[^a-z]", with: "", options: .regularExpression) {
            case "send": return .send
            case "prefill": return .prefill
            case "camera": return .camera
            default: return nil
            }
        }
    }

    private static func inferActions(_ message: String) -> [AgentButtonAction] {
        let lower = message.lowercased()
        if lower.range(of: "\\b(scan|photo|video|camera|snap|picture|image)\\b", options: .regularExpression) != nil {
            return [.camera]
        }
        if message.trimmingCharacters(in: .whitespacesAndNewlines).hasSuffix(":") {
            return [.prefill]
        }
        return [.send]
    }

    // MARK: - Prose

    static func prose(_ content: String) -> String {
        var s = content
        s = replacing("\\[buttons\\][\\s\\S]*?\\[/buttons\\]", in: s, with: "")
        s = replacing("\\[img:[^\\]]+\\]", in: s, with: "")
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Regex helpers

    private static func firstGroup(pattern: String, in s: String) -> String? {
        guard let re = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return nil }
        let ns = s as NSString
        guard let m = re.firstMatch(in: s, range: NSRange(location: 0, length: ns.length)), m.numberOfRanges >= 2 else {
            return nil
        }
        return ns.substring(with: m.range(at: 1))
    }

    private static func replacing(_ pattern: String, in s: String, with t: String) -> String {
        guard let re = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return s }
        let ns = s as NSString
        return re.stringByReplacingMatches(in: s, range: NSRange(location: 0, length: ns.length), withTemplate: t)
    }
}

File: MoveTrack-Android/app/src/main/java/dev/we3kings/nexusmoves/ui/chat/AgentMessageParser.kt
package dev.we3kings.nexusmoves.ui.chat

import java.util.UUID

/**
 * Port of iOS AgentMessageParser. Parses the agent's inline
 * `[BUTTONS]…[/BUTTONS]` blocks into tappable buttons and strips them (plus
 * `[IMG:…]` tokens) from the displayed prose. Mirrors the web contract in
 * movetrack-app NexusChat.vue:
 *
 *   [BUTTONS]
 *   Label|message to send|actions
 *   [/BUTTONS]
 *
 * actions ∈ {send, prefill, camera} (optional; inferred when omitted).
 */
enum class AgentButtonAction { Send, Prefill, Camera }

data class AgentButton(
    val label: String,
    val message: String,
    val actions: List<AgentButtonAction>,
    val id: String = UUID.randomUUID().toString(),
) {
    /** Single acted-on action (priority: prefill > camera > send), matching the web. */
    val primaryAction: AgentButtonAction
        get() = when {
            actions.contains(AgentButtonAction.Prefill) -> AgentButtonAction.Prefill
            actions.contains(AgentButtonAction.Camera) -> AgentButtonAction.Camera
            else -> AgentButtonAction.Send
        }
}

data class ParsedAgentMessage(val prose: String, val buttons: List<AgentButton>)

object AgentMessageParser {
    private val buttonsBlock = Regex("\\[buttons\\]([\\s\\S]*?)\\[/buttons\\]", RegexOption.IGNORE_CASE)
    private val imgToken = Regex("\\[img:[^\\]]+\\]", RegexOption.IGNORE_CASE)
    private val actionsShape = Regex("^[a-z, ]+$", RegexOption.IGNORE_CASE)
    private val cameraWords = Regex("\\b(scan|photo|video|camera|snap|picture|image)\\b", RegexOption.IGNORE_CASE)
    private val nonAlpha = Regex("[^a-z]")

    fun parse(content: String): ParsedAgentMessage =
        ParsedAgentMessage(prose = prose(content), buttons = buttons(content))

    fun buttons(content: String): List<AgentButton> {
        val inner = buttonsBlock.find(content)?.groupValues?.getOrNull(1) ?: return emptyList()
        return inner.split("\n").mapNotNull { rawLine ->
            val line = rawLine.trim()
            if (!line.contains("|")) return@mapNotNull null
            val parts = line.split("|")
            val label = parts[0].trim()
            val lastPart = parts.last().trim()
            val looksLikeActions = parts.size >= 3 &&
                lastPart.length < 30 &&
                actionsShape.matches(lastPart)

            val messageParts = if (looksLikeActions) parts.subList(1, parts.size - 1) else parts.subList(1, parts.size)
            val message = messageParts.joinToString("|").trim()
            val explicit = if (looksLikeActions) parseActions(lastPart) else emptyList()
            val actions = explicit.ifEmpty { inferActions(message) }

            if (label.isEmpty() || message.isEmpty()) null
            else AgentButton(label = label, message = message, actions = actions)
        }
    }

    private fun parseActions(raw: String): List<AgentButtonAction> =
        raw.lowercase().split(',', ' ').mapNotNull { token ->
            when (nonAlpha.replace(token, "")) {
                "send" -> AgentButtonAction.Send
                "prefill" -> AgentButtonAction.Prefill
                "camera" -> AgentButtonAction.Camera
                else -> null
            }
        }

    private fun inferActions(message: String): List<AgentButtonAction> {
        if (cameraWords.containsMatchIn(message)) return listOf(AgentButtonAction.Camera)
        if (message.trim().endsWith(":")) return listOf(AgentButtonAction.Prefill)
        return listOf(AgentButtonAction.Send)
    }

    fun prose(content: String): String {
        var s = content
        s = buttonsBlock.replace(s, "")
        s = imgToken.replace(s, "")
        return s.trim()
    }
}
